oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -312,20 +323,36 @@ impl RuntimeRef {
             return Err(OakStatus::ErrTerminated);
         }
 
-        let mut node_threads = self.node_threads.lock().unwrap();
+        let reference = NodeRef(self.next_node.fetch_add(1, SeqCst));
+        let mut nodes = self.nodes.lock().unwrap();
 
-        if self.is_terminating() {
-            return Err(OakStatus::ErrTerminated);
-        }
-
-        let join_handle = self
+        match self
             .configurations
             .get(module_name)
             .ok_or(OakStatus::ErrInvalidArgs)
             .and_then(|conf| {
-                conf.new_instance(module_name, self.clone(), entrypoint.to_owned(), reader)
-            })?;
-        node_threads.push(join_handle);
+                conf.new_instance(
+                    module_name,
+                    self.clone(),
+                    reference,
+                    entrypoint.to_owned(),
+                    reader,
+                )
+            }) {
+            Ok(join_handle) => {
+                nodes.insert(
+                    reference,
+                    Node {
```

Comment:
Maybe `Configuration::new_instance` should return a `Node` struct directly?

Reply:
Currently the implementation of `Node` is private to the runtime, and would require being visible to the crate for this. I think the encapsulation to only the runtime is preferable, but we can change it in future PRs as necessary.